webbpsf/opds.py

```diff
@@ -1349,15 +1369,100 @@ def _get_basis(*args, **kwargs):
+        ----------
+        Zernike or Hexike coefficients, in microns, for the first nine (9) coefficients, with PTT explicitly set to zero.
+
+        '''
+        
+        # GET SM POSE:
+        # RE-ORDER SUCH THAT (1) X-TRANS, (2) Y-TRANS, (3) X-TILT, (4) Y-TILT.
+        sm_errors = [self.segment_state[-1, 2],    # X-TRANS
+                     self.segment_state[-1, 3],    # Y-TRANS
+                     self.segment_state[-1, 0],    # X-TILT
+                     self.segment_state[-1, 1],    # Y-TILT
+                     self.segment_state[-1, 4]]    # PISTON
+        
+        # GET SM INFLUENCE MATRIX:   
+        if zernike:
+            #ZERNIKE PROJECTED ONTO EXIT PUPIL:
+            print("Beware: using exit pupil Zernike's")
+            smif = astropy.table.Table.read(os.path.join(__location__, 'otelm', "SMIF_zernike.csv"))
+        else:
+            # HEXIKE PROJECTED ONTO ENTRANCE PUPIL (WHAT WEBBPSF NEEDS):
+            smif = astropy.table.Table.read(os.path.join(__location__, 'otelm', "SMIF_hexike.csv"))
+     
+        alphas = smif[smif['Type'] == 'alpha']
+        betas  = smif[smif['Type'] == 'beta']  
+
+        # NOW, WE SUM UP THE ALPHAS AND BETAS FOR EACH HEXIKE/ZERNIKE,
+        # AS DESCRIBED IN THE DOC STRING ABOVE.
+        coeffs = np.zeros((6))
+        for i, icol in enumerate(smif.colnames[2:]):
+            coeffs[i] = np.sum( (alphas[icol]*dx + betas[icol]*dy )*sm_errors)
+
+        if zernike:
+            # SWAP COLUMNS FOR 45-ASTIG AND 0-ASTIG SINCE POPPY ZERNIKE RETURNS 45-ASTIG FIRST
+            coeffs[ [1, 2] ] = coeffs[ [2, 1] ]
+
+        # PAD IN ZEROS FOR PISTON, TIP, TILT:
+        coeffs = np.insert(coeffs, 0, [0., 0., 0.])
+    
+        return coeffs
+        
+        
+    def _apply_sm_field_dependence_model(self, **kwargs):
+        """
+        Apply SM field-dependent model for OTE wavefront error as a function of field angle and SM pose amplitude.
+
+        Updates self.opd based on V2V3 coordinates and amplitude of SM pose.
         """
-        # To be written! Probably will look a lot like _apply_global_zernikes, after
-        # computing the Zernike coefficients based on the multifield model
-        pass
 
+        if self.v2v3 is None:
+            return
+        else:
+            dx =-(self.v2v3[0] - self.ote_control_point[0]).to(u.rad).value # NEGATIVE SIGN B/C TELFER'S FIELD ANGLE COORD. SYSTEM IS (X,Y) = (-V2,V3)
+            dy = (self.v2v3[1] - self.ote_control_point[1]).to(u.rad).value
+
+            z_coeffs = self._get_zernike_coeffs_from_smif(dx, dy, **kwargs)
+
+            perturbation =  poppy.zernike.opd_from_zernikes(z_coeffs, npix=1024, 
+                                                    basis=poppy.zernike.hexike_basis_wss, aperture=self.amplitude)
+
+            perturbation[~np.isfinite(perturbation)] = 0.0
+
+            self.opd += -perturbation*1e-6
+            
+            self.opd_header['SM_IFM'] = 'Applied SMIF field-dependent aberrations'
```

Comment:
minor, this one is probably better suited to be a HISTORY keyword. We might also consider including here in the header some of the numerical values for dx, dy, control point, and so on. (low priority)

Reply:
Done.